app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt

```diff
@@ -0,0 +1,89 @@
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.sci.recipeandroid.feature.auth.data.service.FacebookAuthenticator
+import com.sci.recipeandroid.feature.auth.data.service.GoogleAuthenticator
+import com.sci.recipeandroid.feature.auth.domain.repository.AuthRepository
+import com.sci.recipeandroid.util.SingleLiveEvent
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class AuthOptionViewModel(
+    private val authRepository: AuthRepository,
+    private val googleAuthenticator: GoogleAuthenticator,
+    private val facebookAuthenticator: FacebookAuthenticator,
+) : ViewModel() {
+    private val _uiEvent = SingleLiveEvent<String>()
+    val uiEvent: LiveData<String> = _uiEvent
+
+    fun googleAuthentication(context: Context) {
+        viewModelScope.launch(Dispatchers.IO) {
+            googleAuthenticator.onAuthenticate(context)
+                .fold(
+                    onSuccess = {
+                        authRepository.googleAuthentication(it)
+                            .fold(
+                                onSuccess = { success: String ->
+                                    _uiEvent.postValue(success)
+                                },
+                                onFailure = { error: Throwable ->
+                                    _uiEvent.postValue(
+                                        error.localizedMessage ?: "An Unknown Error Occurred"
+                                    )
+                                }
+                            )
+                    },
+                    onFailure = { error:Throwable ->
+                        if (error !is GetCredentialCancellationException) {
+                            _uiEvent.postValue(
+                                error.localizedMessage
+                                    ?: "An Unknown Error Occurred"
+                            )
+                        }
+
+                    }
+                )
+        }
+    }
+
+    fun facebookLogin(fragment: Fragment) {
+        facebookAuthenticator.facebookLogin(fragment = fragment,
+            onSuccess = { accessToken ->
+                facebookAuthentication(accessToken)
+            },
+            onCancel = {
+                _uiEvent.postValue("Facebook Authentication Cancelled")
+            },
+            onError = {
+                _uiEvent.postValue(it.localizedMessage ?: "Something went wrong")
+            })
+    }
+
+    fun handleFacebookActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        facebookAuthenticator.handleFacebookActivityResult(
+            resultCode = resultCode,
+            requestCode = requestCode,
+            data = data
+        )
+
+    }
+    private fun facebookAuthentication(accessToken: String) {
+        viewModelScope.launch(Dispatchers.IO) {
```

Comment:
plz move IO handling to corresponding layer

Reply:
move io handling to datasource 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c